br/pkg/storage/s3.go

```diff
@@ -954,29 +969,82 @@ func (r *s3ObjectReader) Read(p []byte) (n int, err error) {
 			zap.Error(err),
 		)
 		// if can retry, reopen a new reader and try read again
-		end := r.rangeInfo.End + 1
-		if end == r.rangeInfo.Size {
-			end = 0
-		}
-		_ = r.reader.Close()
-
-		newReader, _, err1 := r.storage.open(r.ctx, r.name, r.pos, end)
-		if err1 != nil {
-			log.Warn("open new s3 reader failed", zap.String("file", r.name), zap.Error(err1))
-			return
-		}
-		r.reader = newReader
-		if r.prefetchSize > 0 {
-			r.reader = prefetch.NewReader(r.reader, r.prefetchSize)
+		if err = r.recreateConn(); err != nil {
+			log.Warn("open new s3 reader failed", zap.String("file", r.name), zap.Error(err))
+			return 0, err
 		}
 		retryCnt++
-		n, err = r.reader.Read(p[:maxCnt])
+		n, err = readFn(p[:maxCnt])
 	}
 
 	r.pos += int64(n)
 	return
 }
 
+func (r *s3ObjectReader) readSlowConn(p []byte) (n int, err error) {
+	// sometimes reading S3 will become very slow when the S3 connection has
+	// no progress for a long time.
+
+	// ~1 MB/s. Even a cold object can provide the throughput of ~18MB/s. Here 1MB/s
+	// threshold is very conservative.
+	expectedBytesPerMilliSec := 1000
+	expectedDuration := time.Duration(len(p)/expectedBytesPerMilliSec) * time.Millisecond
+	// leave some room for RTT. And in the loop, we will double the expectedDuration.
+	expectedDuration = max(expectedDuration, 10*time.Millisecond)
+	for {
+		done := make(chan struct{})
+		go func() {
+			n, err = r.reader.Read(p)
+			close(done)
+		}()
+		select {
+		case <-done:
+			return
+		case <-time.After(expectedDuration):
+			log.Warn("read s3 object is slow, will retry",
+				zap.String("file", r.name),
+				zap.Int64("pos", r.pos),
+				zap.Duration("expectedDuration", expectedDuration))
+			// wait above goroutine exit
+			r.readerCancel()
+			select {
+			case <-r.parentCtx.Done():
+				return 0, r.parentCtx.Err()
+			case <-done:
+			}
+
+			if err = r.recreateConn(); err != nil {
```

Comment:
it might keep recreating if new connection is also slow

Reply:
It's expected. If new connection is slow, we still can't make progress, so still recreating to see if the network is getting better. Also I will double `expectedDuration` to avoid the duration is set too small.